### PR TITLE
[memory_checker] Do not check memory usage of containers which are not created

### PR DESCRIPTION
**Pull request by yozhao101**
Signed-off-by: Yong Zhao <yozhao@microsoft.com>

#### Why I did it
This PR aims to fix an issue (https://github.com/Azure/sonic-buildimage/issues/10088) by enhancing the script `memory_checker`. 

Specifically, if container is not created successfully during device is booted/rebooted, then `memory_checker` do not need check its memory usage.

#### How I did it
In the script `memory_checker`, a function is added to get names of running containers. If the specified container name is not in current running container list, then this script will exit without checking its memory usage.

#### How to verify it
I tested on a lab device by following the steps:

1. Stops `telemetry` container with command `sudo systemctl stop telemetry.service`
2. Removes `telemetry` container with command `docker rm telemetry`
3. Checks whether the script `memory_checker` ran by Monit will generate the syslog message saying it will exit without checking memory usage of `telemetry`:
        
        Jun 14 15:42:08.580612 str-s6000-on-2 INFO /memory_checker: [memory_checker] Exits without checking memory usage since container 'telemetry' is not running!

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [x ] 201911
- [x ] 202006
- [ x] 202012
- [x ] 202106
- [x ] 202111

### CODE REVIEW

**Comment by yozhao101** [comment]
@qiluo-msft Can you help me review this PR please?

File: files/image_config/monit/memory_checker [hunk 1]
```diff
@@ -104,7 +131,13 @@ def main():
     parser.add_argument("threshold_value", type=int, help="threshold value in bytes")
     args = parser.parse_args()
 
-    check_memory_usage(args.container_name, args.threshold_value)
+    running_container_names = get_running_container_names()
+    if args.container_name in running_container_names:
```
Comment (qiluo-msft) [inline files/image_config/monit/memory_checker:135]: >running_container_names [](http://example.com/codeflow?start=30&length=23)

Add a sonic-mgmt test case?

File: files/image_config/monit/memory_checker [hunk 2]
```diff
@@ -94,6 +96,31 @@ def check_memory_usage(container_name, threshold_value):
...
+    """Retrieves names of running containers by talking to the docker daemon.
+
+    Args:
+        None.
+
+    Returns:
+        running_container_names: A list indicates names of running containers.
+    """
+    running_container_names = []
+    docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
```
Comment (nazariig) [inline files/image_config/monit/memory_checker:109]: @yozhao101 do we need handle the case when docker engine is not started? Let's say memory checker starts simultaneously with docker service

Comment (yozhao101) [reply files/image_config/monit/memory_checker:109]: `memory_checker` script will be ran by Monit periodically and Monit is delayed to run this script after device is up 300 seconds. 

Docker daemon will be started by systemd immediately once device is booted/rebooted.

If docker daemon crashed or is not started successfully, then `memory_checker` will exit and log an error message into syslog.

File: files/image_config/monit/memory_checker (hunk 1 above)
Comment (yozhao101) [reply files/image_config/monit/memory_checker:135]: Working on the test case and will post the PR link at here.

**Comment by yozhao101** [comment]
@nazariig Can you please review again to see whether logging an error message is good way or not to address your question?

File: files/image_config/monit/memory_checker [hunk 3]
```diff
@@ -94,6 +96,31 @@ def check_memory_usage(container_name, threshold_value):
...
+        running_container_names: A list indicates names of running containers.
+    """
+    running_container_names = []
+    try:
+        docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
+        container_obj = docker_client.containers
+        container_list = container_obj.list(filters={"status": "running"})
+        for container in container_list:
+            running_container_names.append(container.name)
+    except (docker.errors.APIError, docker.errors.DockerException) as err:
```
Comment (nazariig) [inline files/image_config/monit/memory_checker:115]: @yozhao101 i would suggest to rewrite like this:
```python
    try:
        docker_client = docker.DockerClient(base_url='unix://var/run/docker.sock')
        running_container_list = docker_client.containers.list(filters={"status": "running"})
        running_container_names = [ container.name for container in running_container_list ]
    except (docker.errors.APIError, docker.errors.DockerException) as err:
```

File: files/image_config/monit/memory_checker (hunk 2 above)
Comment (nazariig) [reply files/image_config/monit/memory_checker]: @yozhao101 ok

File: files/image_config/monit/memory_checker (hunk 3 above)
Comment (yozhao101) [reply files/image_config/monit/memory_checker:112]: Updated.

File: files/image_config/monit/memory_checker (hunk 1 above)
Comment (yozhao101) [reply files/image_config/monit/memory_checker:132]: sonic-mgmt test case is posted: https://github.com/Azure/sonic-mgmt/pull/5823. If you are available, can you please help me review?

**Review by qiluo-msft — COMMENTED** [review]
LGTM. Please wait for other reviewers.

**Review by nazariig — APPROVED** [review]

**Comment by yozhao101** [comment]
@qiluo-msft another reviewer approved this PR, can you also approve it please if there is no more comments?

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**